cli/command/manifest/push.go

```diff
@@ -0,0 +1,591 @@
+	}
+	pushURL, err := createManifestURLFromRef(targetRef, urlBuilder)
+	if err != nil {
+		return errors.Wrapf(err, "error setting up repository endpoint and references for %q: %v", targetRef)
+	}
+	logrus.Debugf("manifest list push url: %s", pushURL)
+
+	deserializedManifestList, err := manifestlist.FromDescriptors(manifestList.Manifests)
+	if err != nil {
+		return errors.Wrap(err, "cannot deserialize manifest list")
+	}
+	mediaType, p, err := deserializedManifestList.Payload()
+	logrus.Debugf("mediaType of manifestList: %s", mediaType)
+	if err != nil {
+		return errors.Wrap(err, "cannot retrieve payload for HTTP PUT of manifest list")
+
+	}
+	putRequest, err := http.NewRequest("PUT", pushURL, bytes.NewReader(p))
+	if err != nil {
+		return errors.Wrap(err, "HTTP PUT request creation failed")
+	}
+	putRequest.Header.Set("Content-Type", mediaType)
+
+	httpClient, err := getHTTPClient(ctx, dockerCli, targetRepo, targetEndpoint, targetRepoName)
+	if err != nil {
+		return errors.Wrap(err, "failed to setup HTTP client to repository")
+	}
+
+	// before we push the manifest list, if we have any blob mount requests, we need
+	// to ask the registry to mount those blobs in our target so they are available
+	// as references
+	if err := mountBlobs(httpClient, urlBuilder, targetRef, blobMountRequests); err != nil {
+		return errors.Wrap(err, "couldn't mount blobs for cross-repository push")
+	}
+
+	// we also must push any manifests that are referenced in the manifest list into
+	// the target namespace
+	// Use the untagged target for this so the digest is used
+	if err := pushReferences(httpClient, urlBuilder, bareRef, manifestRequests); err != nil {
+		return errors.Wrap(err, "couldn't push manifests referenced in our manifest list")
+	}
+
+	resp, err := httpClient.Do(putRequest)
+	if err != nil {
+		return errors.Wrap(err, "v2 registry PUT of manifest list failed")
+	}
+	defer resp.Body.Close()
+
+	if statusSuccess(resp.StatusCode) {
+		dgstHeader := resp.Header.Get("Docker-Content-Digest")
+		dgst, err := digest.Parse(dgstHeader)
+		if err != nil {
+			return err
+		}
+		if opts.purge == true {
+			logrus.Debugf("deleting files at %s", targetFilename)
+			if err := os.RemoveAll(targetFilename); err != nil {
+				// Not a fatal error
+				logrus.Info("unable to clean up manifest files in %s", targetFilename)
+			}
+		}
+		logrus.Infof("succesfully pushed manifest list %s with digest %s", targetRef, dgst)
+		return nil
+	}
+	return fmt.Errorf("registry push unsuccessful: response %d: %s", resp.StatusCode, resp.Status)
+}
+
+func getYamlInput(dockerCli *command.DockerCli, yamlFile string) (YamlInput, error) {
+	logrus.Debugf("YAML file: %s", yamlFile)
+
+	if _, err := os.Stat(yamlFile); err != nil {
+		logrus.Debugf("unable to open file: %s", yamlFile)
+	}
+
+	var yamlInput YamlInput
+	yamlBuf, err := ioutil.ReadFile(yamlFile)
+	if err != nil {
+		logrus.Fatalf(fmt.Sprintf("can't read YAML file %q: %v", yamlFile, err))
```

Comment:
I don't think we should be using `Fatalf()` at all. We should return the errors here (and below). Otherwise we miss out on the change to wrap the error with important context.